Context manager support for Services

bpkroth opened the pull request:
Work towards making the setup/teardown work for the SshService managed by a Context instead of global.

To do this, we introspect all of the methods registered for a Service mixin (which we also renamed from `self._services` to `self._service_methods`), and pull out their unique Service instance references from the `__self__` attribute in the bound method that was registered, then store them as `self._services`.
This is to handle the case where a single method name was overridden multiple times.
At this point, we have enough information to invoke a sort of `__enter__` context for each of those Service instances.
However, since we use `__enter__` to invoke all of them, we instead need to implement a separate `_enter_context()` method that does the actual work, and can then be overridden by the subclasses.
 
Unfortunately, this technique make tamper with the order the Services were layered on one another, so any logic that required relying on that order for context setup might be problematic.  I don't know if that will actually be an issue, so for now, I'm ignoring the issue and just documenting it.

This PR does not include explicit new tests for this, though does have to fixup a few mocks in order to make sure that we're registered bound methods and not just functions.

#510 will add additional tests after this is merged.

motus reviewed (approved):
ok let's merge it in and thing of a more generic solution later. in particular, I would love to add back the support for arbitrary static functions and lambdas as service methods

bpkroth commented on mlos_bench/mlos_bench/services/base_service.py:
```diff
@@ -79,13 +81,34 @@ def __init__(self,
+        # To do this we also
+
+        # As a sanity check, make sure all methods registered are bound.
+        assert all(hasattr(svc_method, '__self__') and isinstance(svc_method.__self__, Service)
+                   for svc_method in self._service_methods.values())
```
Yeah, I'd already adjusted those to make them methods, but it is a reasonable point.
Initially I was thinking to use that as a sanity test to make sure I wasn't missing something.
But if we want I could use it as a filter instead so that only service instances are considered for context entry.
I'm not sure there'd be a use case for non class instances for that.

bpkroth replied:
Reverted that and changed it to just a filter.

bpkroth commented:
> ok let's merge it in and think of a more generic solution later. in particular, I would love to add back the support for arbitrary static functions and lambdas as service methods

I reverted that part and added support for static methods back in.

Just means they can't participate in Service content enter/exit.
